File: Recursion/binary_search.c
/* Searching element in sorted array using binary search */

#include <stdio.h>

int search(int *arr, int low, int high, int target)
{
    if (high > low)
    {
        int mid = (high - low) / 2;
        if (arr[mid] == target)
        {
            return mid;
        }
        else
        {
            if (arr[mid] < target)
            {
                return search(arr, mid + 1, high, target);
            }
            else
            {
                return search(arr, low, mid - 1, target);
            }
        }
    }
    return -1;
}

int main()
{
    int arr[] = {1, 2, 3, 4, 6, 8, 9};
    int size = sizeof(arr) / sizeof(int);
    int result = search(arr, 0, size, 2);
    if (result > 0)
    {
        printf("Target found at index %d", result);
    }
    else
    {
        printf("Target not found!");
    }

    return 0;
}

File: Searching/binary_search.py
"""
Binary Search quick revision notes:
- Works only on sorted data.
- Compare target with middle element, then discard half of the search space.
- Repeat until found or range becomes empty.
- Time: O(log n)
- Space: O(1) auxiliary (iterative version)
"""


def binary_search(elements, target):
    """Return target index if found, else False."""
    low = 0
    high = len(elements) - 1

    while low <= high:
        mid = low + (high - low) // 2
        if elements[mid] == target:
            return mid
        elif elements[mid] > target:
            high = mid - 1
        elif elements[mid] < target:
            low = mid + 1

    return False


if __name__ == "__main__":
    elements = [5, 7, 9, 12, 32, 59]
    target = 32
    index = binary_search(elements, target)

    # Use explicit check so index 0 is treated as a valid match.
    if index is not False:
        print(f"Element '{target}' found at index {index}")
    else:
        print("Not Found")
